From 08f0573ed02184e4940887e523cbea191102bc5c Mon Sep 17 00:00:00 2001
From: Thiago Macieira <thiago.macieira@intel.com>
Date: Tue, 16 Jul 2019 12:35:04 -0700
Subject: [PATCH] Make updatedb create the DB dir

We could create it in the .service file (StateDirectory=mlocate), but
that wouldn't help running updatedb from the command-line for the
first time.

Signed-off-by: Thiago Macieira <thiago.macieira@intel.com>
---
 src/updatedb.c | 40 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 40 insertions(+)

diff --git a/src/updatedb.c b/src/updatedb.c
index 7cfa22d..3614bba 100644
--- a/src/updatedb.c
+++ b/src/updatedb.c
@@ -895,6 +895,45 @@ unlink_init (void)
     sigaction (signals[i], &sa, NULL);
 }
 
+/* Create the directory where the database will reside, if it doesn't
+   exist yet. Exit on error. */
+static void
+create_db_dir (void)
+{
+  char *dirname = xstrdup (conf_output);
+  char *slash = strrchr (dirname, '/');
+  int ret;
+  gid_t gid = -1;
+
+  if (slash == NULL)
+    return;                     /* relative path */
+  *slash = '\0';
+
+  if (conf_check_visibility != false)
+    {
+      struct group *grp;
+
+      grp = getgrnam (GROUPNAME);
+      if (grp == NULL)
+	error (EXIT_FAILURE, errno, _("can not find group `%s'"), GROUPNAME);
+      gid = grp->gr_gid;
+    }
+
+  ret = mkdir (dirname, S_IRUSR | S_IWUSR | S_IXUSR | S_IRGRP | S_IXGRP);
+  if (ret != 0 && errno != EEXIST)
+    error (EXIT_FAILURE, errno, _("can not create directory `%s'"),
+           dirname);
+  if (ret == 0 && gid != -1)
+    {
+      if (chown (dirname, (uid_t)-1, gid) != 0)
+	error (EXIT_FAILURE, errno,
+	       _("can not change group of directory `%s' to `%s'"), dirname,
+	       GROUPNAME);
+    }
+
+  free (dirname);
+}
+
  /* Top level */
 
 /* Open a temporary file for the new database and initialize its header
@@ -1000,6 +1039,7 @@ main (int argc, char *argv[])
 	}
     }
   unlink_init ();
+  create_db_dir ();
   new_db_open ();
   dir_state_init (&scan_dir_state);
   if (chdir (conf_scan_root) != 0)
-- 
2.22.0

